Client: Java ID generator helper

kprotty opened the pull request:
Creates a static helper function in the java client for generating thread-safe, unique, and monotonic `u128` IDs to use for Accounts and Transfers.

kprotty committed:
java: Uint128.ULID (wip)

kprotty committed:
java: fix ULID monotonic/overflow detection & tests

kprotty committed:
java: fix ULID test + add thread test

The ULID spec notes that IDs within the same millisecond dont need to be
ordered: https://github.com/ulid/spec?tab=readme-ov-file#sorting

kprotty committed:
java: ULID random overflow from absolute increment

Overflow should happen relative to the 80-bit random generated at the
timestamp rather than relative to a separate counter at the timestamp
which is applied to the random.

This provides sort order guarantee using the recommended approach from
the spec: https://github.com/ulid/spec?tab=readme-ov-file#monotonicity

batiati commented on src/clients/java/src/main/java/com/tigerbeetle/UInt128.java:
```diff
@@ -172,5 +174,53 @@ public static byte[] asBytes(final BigInteger value) {
+
+            // Increment lo. If that overflows, increment hi. If that overflows, throw error.
+            // ULID spec says to throw error upon 80-bit overflow (not relative to initial random).
+            if (randomLo++ == Long.MAX_VALUE && randomHi++ == 0xffff) {
```
Zig has spoiled me a lot 😇

- Can we drop the `++` operator in favor of `+=1`?

- It's a signed integer, did you really mean `Long.MAX_VALUE` here, or should it be `0xFFFFFFFFFFFFFFFF` ?

- It's safe to use `+ 1` with signed integers since Java allows arithmetic operations to overflow. But it's worthy of a comment.

batiati commented on src/clients/java/src/test/java/com/tigerbeetle/UInt128Test.java:
```diff
@@ -165,4 +169,89 @@ public void testAsBigIntegerZero() {
+                        }
+
+                        var ulidB = new DecodedULID(UInt128.ULID());
+                        ulidB.assertMonotonicSince(ulidA);
```
Can we also use `UInt128.asBigInteger` to compare both ULIDs?
This will assert that the binary representation is really what we expect at Zig's side (a monotonically increasing `u128` little-endian).

```java
var bigIntA = UInt128.asBigInteger(...);
var bigIntB = UInt128.asBigInteger(...);
assertTrue(bigIntB.compareTo(bigIntA) > 0); 
```

kprotty replied:
Sort of: `asBigInteger` interprets the bytes as 2 little-endian u64 values of a u128. ULID however stores timestamp as `{u32, u16}` and random as u80, both big-endian. We can instead interpret the bytes directly via `new BigInteger(ULID())` since it assumes big-endian. This also means Zig side might not see it as monotonically increasing little-endian.

kprotty replied:
`ULID` has been renamed to `ID` and it stores the individual fields in little-endian which allows use of `asBigInteger`.

batiati reviewed (commented):
Nice and elegant implementation!!!

Can we push the scope of this PR a little further and review the samples and docs to mention/use ULID where it's applicable?

batiati commented on src/clients/java/src/main/java/com/tigerbeetle/UInt128.java:
```diff
@@ -172,5 +174,53 @@ public static byte[] asBytes(final BigInteger value) {
+     *
+     * @return an array of 16 bytes representing the 128-bit value.
+     */
+    public static byte[] ULID() throws ArithmeticException {
```
By convention unchecked exceptions should not be declared in the `throws` clause, but only documented.

```suggestion
    /**
     * Generates a Universally Unique Lexicographically Sortable Identifier
     * (<a href="https://github.com/ulid/spec">ULID</a>) as 16 bytes of a 128-bit value.
     *
     * The ULID returned always increases monotonically and is thread-safe.
     *
     * @throws ArithmeticException as per ULID specs, if the 80 bits of randomness overflows.
     * @return an array of 16 bytes representing the 128-bit value.
     */
    public static byte[] ULID() {
```

kprotty committed:
java: explicitly use `zig.exe` when on windows

For setups which download both `zig/zig` and `zig/zig.exe` for convenience
(i.e. windows WSL), this ensures the Java CI uses the correct one instead of
relying on shell to append `.exe` (which doesn't happen since non `.exe`
version exists).

kprotty committed:
java: cleanup ULID overflow & comparison

kprotty committed:
java: switch from ULID to unique little-endian IDs

ULID spec requires storing the values as big endian which means, when interpreted
by a replica on the Zig side, the `u128` ID is no longer monotonically incrementing.
This resorts in unique but effectively randomly-ordered IDs which doesn't take
advantage of data/lookup optimizations in the storage engine.

Instead, the Java client now exposes a generic `ID()` helper (instead of specifically `ULID`)
which allows us to return u128 bytes in little-endian format while still using the core
ideas around uniqueness & monotonicity from ULID underneath.

kprotty committed:
java: Use contiguous 48bit timestamp & update docs

Switches to using the 48 LSB of timestamp instead of 16 LSB and 32 MSB. Also updates the data-modeling
documentation to give note about storing ULID fields in little endian.

batiati commented on src/clients/java/src/main/java/com/tigerbeetle/UInt128.java:
```diff
@@ -172,5 +174,60 @@ public static byte[] asBytes(final BigInteger value) {
+            // If randomLo will overflow from increment, then increment randomHi as carry.
+            // If randomHi will overflow on increment, throw error on 80-bit random overflow.
+            if (randomLo == 0xFFFFFFFFFFFFFFFFL) {
+                if (randomHi == 0xffff) {
```
Let's standardize hex in upper case.

```suggestion
                if (randomHi == 0xFFFF) {
```

batiati commented on src/clients/java/src/main/java/com/tigerbeetle/UInt128.java:
```diff
@@ -172,5 +174,60 @@ public static byte[] asBytes(final BigInteger value) {
+    /**
+     * Generates a Universally Unique Sortable Identifier as 16 bytes of a 128-bit value.
+     *
+     * The ID() function is thread-safe, the bytes returned are formatted in little endian, and the
```
I think it's worth mentioning that this ID is based on the UILD spec, except it is little-endian, so it is binary sortable and not lexicographically sortable by the string representation. In fact, our version has no concept of string representation.

```suggestion
     * The ID() function is thread-safe, the bytes returned are stored in little endian, and the
```

batiati reviewed (changes requested):
Just two minor comments, but LGTM other than that!

kprotty committed:
java: cleanup ID() comments & literals

Standardize upper-case hex literals and refer to ULID spec as `ID()`'s inspiration.

batiati reviewed (dismissed):
Thanks King ✌️

kprotty committed:
java: mvn format
